core/src/Kokkos_SYCL_Space.hpp

```diff
@@ -62,15 +63,132 @@ class SYCLDeviceUSMSpace {
   SYCLDeviceUSMSpace();
 
   void* allocate(const std::size_t arg_alloc_size) const;
+  void* allocate(const char* arg_label, const size_t arg_alloc_size,
+                 const size_t arg_logical_size = 0) const;
+
   void deallocate(void* const arg_alloc_ptr,
                   const std::size_t arg_alloc_size) const;
+  void deallocate(const char* arg_label, void* const arg_alloc_ptr,
+                  const size_t arg_alloc_size,
+                  const size_t arg_logical_size = 0) const;
 
   static constexpr const char* name() { return "SYCLDeviceUSM"; };
 
  private:
   int m_device;
 };
 }  // namespace Experimental
+
+namespace Impl {
+
+void Default_USM_memcpy(void* dst, const void* src, size_t n);
+
+template <>
+struct DeepCopy<Kokkos::Experimental::SYCLDeviceUSMSpace,
+                Kokkos::Experimental::SYCLDeviceUSMSpace,
+                Kokkos::Experimental::SYCL> {
+  DeepCopy(void* dst, const void* src, size_t);
+  DeepCopy(const Kokkos::Experimental::SYCL&, void* dst, const void* src,
+           size_t);
+};
+
+template <>
+struct DeepCopy<Kokkos::HostSpace, Kokkos::Experimental::SYCLDeviceUSMSpace,
+                Kokkos::Experimental::SYCL> {
+  DeepCopy(void* dst, const void* src, size_t);
+  DeepCopy(const Kokkos::Experimental::SYCL&, void* dst, const void* src,
+           size_t);
+};
+
+template <>
+struct DeepCopy<Kokkos::Experimental::SYCLDeviceUSMSpace, Kokkos::HostSpace,
+                Kokkos::Experimental::SYCL> {
+  DeepCopy(void* dst, const void* src, size_t);
+  DeepCopy(const Kokkos::Experimental::SYCL&, void* dst, const void* src,
+           size_t);
+};
+
+template <class ExecutionSpace>
+struct DeepCopy<Kokkos::HostSpace, Kokkos::Experimental::SYCLDeviceUSMSpace,
+                ExecutionSpace> {
+  DeepCopy(void* dst, const void* src, size_t n) {
+    (void)DeepCopy<Kokkos::HostSpace, Kokkos::Experimental::SYCLDeviceUSMSpace,
+                   Kokkos::Experimental::SYCL>(dst, src, n);
+  }
+
+  DeepCopy(const ExecutionSpace& exec, void* dst, const void* src, size_t n) {
+    exec.fence();
+    Default_USM_memcpy(dst, src, n);
```

Comment:
Isn't it equivalent to `DeepCopy<HostSpace,SYCLDeviceUSMSpace,SYCL>(SYCL(), dest, src, n);` in which case you wouldn't expose implementation details?

Reply:
The only difference I see is that we fence on the `HostSpace` instance (if `ExecutionSpace=HostSpace`) instead of the default SYCL instance before calling the `SYCL` `memcpy` function. Correct me if I'm wrong.
I tried to mirror what we do for `HIP` here.